File: engine/src/theseus_engine/serialize.py
"""Grid snapshots: save / load / diff — the seed of the M2 sidecar format.

What persists is the MAP, not the sensor history: derived cell states and
semantic labels round-trip exactly; log-odds are collapsed to each
state's saturation value on load (identical to set_state, which the
simulators use too). Run-length encoding keeps a 19k-cell studio under a
few KB of JSON.

diff() is the "what changed since yesterday" kernel: load yesterday's
sidecar, diff against today's model, get labeled changes. On-device this
becomes per-chunk hashes so unchanged chunks never load — but the
semantics are frozen here first.
"""

from __future__ import annotations

import json
from pathlib import Path

from .geometry import Cell
from .grid import OccupancyGrid, UNKNOWN

SCHEMA = "theseus-grid/1"


def to_dict(grid: OccupancyGrid) -> dict:
    states: list[int] = []
    run_state = grid.state((0, 0))
    run_len = 0
    for y in range(grid.height):
        for x in range(grid.width):
            s = grid.state((x, y))
            if s == run_state:
                run_len += 1
            else:
                states.extend((run_state, run_len))
                run_state, run_len = s, 1
    states.extend((run_state, run_len))
    return {
        "schema": SCHEMA,
        "w": grid.width,
        "h": grid.height,
        "cell": grid.cell_size,
        "origin": [grid.origin[0], grid.origin[1]],
        "states": states,
        "labels": {str(i): lab for i, lab in sorted(grid.labels.items())},
    }


def from_dict(d: dict) -> OccupancyGrid:
    if d.get("schema") != SCHEMA:
        raise ValueError(f"unsupported grid schema: {d.get('schema')!r}")
    g = OccupancyGrid(d["w"], d["h"], d["cell"],
                      (d["origin"][0], d["origin"][1]))
    i = 0
    rle = d["states"]
    for k in range(0, len(rle), 2):
        state, run = rle[k], rle[k + 1]
        if state != UNKNOWN:  # grids start all-UNKNOWN
            for j in range(i, i + run):
                g.set_state((j % g.width, j // g.width), state)
        i += run
    if i != g.width * g.height:
        raise ValueError("RLE state stream does not match grid size")
    for key, lab in d.get("labels", {}).items():
        g.labels[int(key)] = lab
    return g


def save(grid: OccupancyGrid, path: str | Path) -> Path:
    p = Path(path)
    p.parent.mkdir(parents=True, exist_ok=True)
    p.write_text(json.dumps(to_dict(grid), separators=(",", ":")),
                 encoding="utf-8")
    return p


def load(path: str | Path) -> OccupancyGrid:
    return from_dict(json.loads(Path(path).read_text(encoding="utf-8")))


def diff(a: OccupancyGrid, b: OccupancyGrid) -> list[tuple[Cell, int, int]]:
    """Cells whose derived state differs: (cell, state_in_a, state_in_b)."""
    if (a.width, a.height) != (b.width, b.height):
        raise ValueError("grids must share dimensions to diff")
    out: list[tuple[Cell, int, int]] = []
    for y in range(a.height):
        for x in range(a.width):
            sa, sb = a.state((x, y)), b.state((x, y))
            if sa != sb:
                out.append(((x, y), sa, sb))
    return out


def diff_report(a: OccupancyGrid, b: OccupancyGrid) -> dict:
    """Human-oriented summary of a->b: what appeared, what vanished,
    grouped by semantic label where one is known."""
    from .grid import OCCUPIED
    appeared = vanished = 0
    by_label: dict[str, int] = {}
    for c, sa, sb in diff(a, b):
        if sb == OCCUPIED and sa != OCCUPIED:
            appeared += 1
        elif sa == OCCUPIED and sb != OCCUPIED:
            vanished += 1
        else:
            continue
        lab = b.label(c) or a.label(c) or "?"
        by_label[lab] = by_label.get(lab, 0) + 1
    return {"appeared": appeared, "vanished": vanished, "by_label": by_label}

File: apps/navcore/Sources/NavCore/Serialize.swift
// Grid snapshots — port of engine serialize.py: save / load / diff, the
// seed of the M2 sidecar format. What persists is the MAP, not sensor
// history: derived states + labels round-trip exactly; log-odds collapse
// to saturation values on load (identical to setState).
//
// Per docs/PORT.md rule 6 the actual JSON bytes live OUTSIDE NavCore
// (Foundation-allowed layers: the app, the replay tests); this file owns
// the semantics — RLE encode/decode and diffing.

public let GRID_SCHEMA = "theseus-grid/1"

public struct GridSnapshot: Sendable {
    public var schema: String
    public var w: Int
    public var h: Int
    public var cell: Double
    public var origin: Vec
    public var states: [Int]          // RLE pairs: state, run, state, run…
    public var labels: [Int: String]
}

public func toSnapshot(_ grid: OccupancyGrid) -> GridSnapshot {
    var states: [Int] = []
    var runState = grid.state(Cell(0, 0))
    var runLen = 0
    for y in 0..<grid.height {
        for x in 0..<grid.width {
            let s = grid.state(Cell(Int32(x), Int32(y)))
            if s == runState {
                runLen += 1
            } else {
                states.append(runState)
                states.append(runLen)
                runState = s
                runLen = 1
            }
        }
    }
    states.append(runState)
    states.append(runLen)
    return GridSnapshot(schema: GRID_SCHEMA, w: grid.width, h: grid.height,
                        cell: grid.cellSize, origin: grid.origin,
                        states: states, labels: grid.labels)
}

public enum SnapshotError: Error {
    case badSchema(String)
    case badLength
}

public func fromSnapshot(_ d: GridSnapshot) throws -> OccupancyGrid {
    guard d.schema == GRID_SCHEMA else {
        throw SnapshotError.badSchema(d.schema)
    }
    let g = OccupancyGrid(width: d.w, height: d.h, cellSize: d.cell,
                          origin: d.origin)
    var i = 0
    var k = 0
    while k < d.states.count {
        let state = d.states[k]
        let run = d.states[k + 1]
        if state != UNKNOWN {  // grids start all-UNKNOWN
            for j in i..<(i + run) {
                g.setState(Cell(Int32(j % g.width), Int32(j / g.width)),
                           state)
            }
        }
        i += run
        k += 2
    }
    guard i == g.width * g.height else {
        throw SnapshotError.badLength
    }
    for (key, lab) in d.labels {
        g.labels[key] = lab
    }
    return g
}

/// Cells whose derived state differs: (cell, state in a, state in b).
public func gridDiff(_ a: OccupancyGrid,
                     _ b: OccupancyGrid) -> [(Cell, Int, Int)] {
    precondition(a.width == b.width && a.height == b.height,
                 "grids must share dimensions to diff")
    var out: [(Cell, Int, Int)] = []
    for y in 0..<a.height {
        for x in 0..<a.width {
            let c = Cell(Int32(x), Int32(y))
            let sa = a.state(c)
            let sb = b.state(c)
            if sa != sb {
                out.append((c, sa, sb))
            }
        }
    }
    return out
}

public struct DiffReport: Sendable, Equatable {
    public var appeared: Int
    public var vanished: Int
    public var byLabel: [String: Int]
}

/// Human-oriented summary of a->b: what appeared, what vanished, grouped
/// by semantic label where one is known.
public func diffReport(_ a: OccupancyGrid, _ b: OccupancyGrid) -> DiffReport {
    var appeared = 0
    var vanished = 0
    var byLabel: [String: Int] = [:]
    for (c, sa, sb) in gridDiff(a, b) {
        if sb == OCCUPIED && sa != OCCUPIED {
            appeared += 1
        } else if sa == OCCUPIED && sb != OCCUPIED {
            vanished += 1
        } else {
            continue
        }
        var lab = b.label(c)
        if lab.isEmpty { lab = a.label(c) }
        if lab.isEmpty { lab = "?" }
        byLabel[lab, default: 0] += 1
    }
    return DiffReport(appeared: appeared, vanished: vanished,
                      byLabel: byLabel)
}
